Add form for reassigning users to places #257

freddieptf opened the pull request:
Part of https://github.com/medic/cht-user-management/issues/185
Fixes #260

This adds the form for re-assigning a user to a place.

freddieptf commented on src/routes/reassign.ts:
```diff
@@ -0,0 +1,61 @@
+    const body =  req.body as { [key:string]:string };
+    const contactType = Config.getContactType(body.place_type);
+
+    const uuidMatch = body.contact.match('/contacts/(?<uuid>[a-z0-9]{32})');
```
The regex is cause of a weird thing i found, place UUIDs had hyphens but people did not. I thought this was consistent behavior but I just checked on a live instance and it seems this is not always the case. I'll update this

freddieptf replied:
> I don't think you should use the string inside the match because that will literally search for the string '/contacts/(?<uuid>[a-z0-9]{32})' instead of using it as a pattern.

The regex worked for me when i had UUIDs without hyphens

Benmuiruri reviewed (commented):
I was getting this error. 

<details>
<summary>Error</summary>
<img width="696" alt="Screenshot 2025-02-28 at 17 55 48" src="https://github.com/user-attachments/assets/c412b8b3-b32c-4e28-b0c5-745bcd483458" />

</details>

But I now see you changed the implementation. Let me continue with the review

Benmuiruri reviewed (commented):
Nice work @freddieptf the reassignment works. 

I have not checked everything but here is one thing I found. When you reassign a place from person A to person B. We change the contact person of the place and replace it with person B. If you want to reassign the place back to person A, how would one go about that? We now cannot find the person. Check attached video 

<details>
<summary>Contact disappears</summary>


https://github.com/user-attachments/assets/ec86cb35-e7c1-4201-a571-c399d6c99a2f



</details>

freddieptf commented:
The behavior you observed was due to the fact that the contact was only supervising one place. If a contact is only supervising one place and the place gets assigned to another person, then their user account gets disabled and their contact gets deleted.

>  If you want to reassign the place back to person A, how would one go about that? 

They'd have to get their contact and user created manually on the instance. This process can be improved in https://github.com/medic/cht-user-management/pull/257

Benmuiruri reviewed (changes requested):
The happy path works to reassign a CHU but I have left some comments and questions that could make the implementation more robust (in terms of error handling) and DRY. 

Suggested improvements 

- When you submit an incorrect uuid the application fails silently by logging the error message but when you click done, the page breaks with `Site cannot be reached` 
- when you submit a correct contact link but that person is a CHP it shows failure 400, that error message should better communicate to the user why it failed (Selected contact is not a CHA)
- when you submit a correct contact link of a person who is not a CHA it fails silently in the terminal but the page breaks and you get `Site cannot be reached`

<details>
<summary>Silent failure</summary>
<img width="1004" alt="Screenshot 2025-03-03 at 09 26 50" src="https://github.com/user-attachments/assets/23b3f3f1-cd60-4845-a632-8d3b559a3d80" />

</details>

- When reassignment is successful, we could improve the feedback to the user with something like `Successfully reassigned place xxx to user yyy` in green. The success is good but it is immediately clear that the process succeeded.

kennsippell commented on src/liquid/reassign/form.liquid:
```diff
@@ -0,0 +1,132 @@
+      </div>
+    </div>
+
+    {% if places and places.length > 0 %}
```
This is another paradigm shift to have the table, staging, and upload progress here on the form instead of on the homepage. I have two broad concerns.

First concern is that I'm unclear how you're going to implement CSV import for CHA reassignment and creation. Will there be another form with another table? Rows don't appear editable presently which is key to CSV import experience. Do warnings work? How do you retry if things fail (I tested and didn't see an error). 

Would you provide a mockup of how you're expecting the top-level nav to work for this?  I think CSV important is really important, so it'd be good to think some of these things through and get on the same page before proceeding too far.

My second concern is just around consistency of the experience. Currently staging and uploading is orchestrated on the home page. I'm not saying what is done here is undesirable, but it is inconsistent.

freddieptf replied:
> Would you provide a mockup of how you're expecting the top-level nav to work for this? I think CSV important is really important, so it'd be good to think some of these things through and get on the same page before proceeding too far.

All the items under `Manage` on the navbar should have a status page/list of their own. We can have tabs on the homepage or different pages for these

> Will there be another form with another table? Rows don't appear editable presently which is key to CSV import experience. Do warnings work?

Yes there will be another form and table reusing the views in this one. I don't think we need warnings/validation as the places are already created.

> How do you retry if things fail (I tested and didn't see an error).

Right now, unfortunately, restart from scratch. This will be improved

freddieptf commented on src/routes/reassign.ts:
```diff
@@ -0,0 +1,126 @@
+  
+  fastify.get('/reassign', async (req, resp) => {
+    const { place_type } = req.query as any;
+    const contactType = Config.getContactType(place_type);
```
Not strictly, the user would have to manually edit the query param in the URL for this to fail

freddieptf commented on src/lib/cht-api.ts:
```diff
@@ -184,6 +193,16 @@ export class ChtApi {
 
+  async getUser(contactId: string): Promise<UserInfo> {
+    const url = `api/v2/users?contact_id=${contactId}`;
+    console.log('axios.get', url);
```
These are actually needed when debugging both in dev and prod, that's why every request in `cht-api` has a log statement

freddieptf commented on src/liquid/reassign/form.liquid:
```diff
@@ -0,0 +1,132 @@
+        >
+          https://{{ session.authInfo.domain -}}
+        </a>
+        then copy and paste their link below
```
> In all cases, the user types the name of whatever contact they are looking for.

I don't feel like we can guarantee a performant search as is, that's why i opted for copying the URL from the CHT. If we can provide the same "general search" experience as the CHT then i think that might be better than copying the URL/UUID especially if we can get the contact hierarchy in the results as well.

freddieptf commented on src/liquid/app/nav.html:
```diff
@@ -35,6 +35,9 @@
         </a>
         <div class="navbar-dropdown">
           {% for placeType in contactTypes %}
+            {% if placeType.can_assign_multiple %}
```
It was never there, you click "Add x" as before and if the instance version is >= 4.9.0 and `can_assign_multiple` is set then it loads the new form

freddieptf commented:
@kennsippell ready for review, I've also updated the replacement functions to take account of users with multiple places.

freddieptf commented:
Pinging for another review @Benmuiruri @kennsippell, curious to see if you think the UX has improved abit

Benmuiruri reviewed (changes requested):
@freddieptf awesome progress. I have 2 concern and 2 suggestions.

**Concern** 
- I was able to reassign a place to a CHA A which belonged to CHA B, but CHA B can still see the place when they login. CHA B can also send reports of the place even though it has been moved to CHA A. When CHA A refreshes they see "Ohh a new report just appeared and I didn't submit this."

 - UI: The top navbar is missing while on the Reassign Community Health Unit page 

**Suggestions**
- It still does not give feedback to users if the contact is not a CHA. If I try to Reassign a CHU to a valid contact who is a CHP, the page refreshes, I don't see the success but at the same time I don't see a fail. Do you think we need to check if the contact has the cha role ? and if they don't then we show an error, the same way we display when contact is not found via an incorrect uuid ? 

 - UI change - The Reassign Community Health Unit page is inconsistent from the other pages like Move Community Health Unit (for instance the header is not bold. It would be nice to have consistent design on the pages)

freddieptf commented:
@Benmuiruri 

> I was able to reassign a place to a CHA A which belonged to CHA B, but CHA B can still see the place when they login. CHA B can also send reports of the place even though it has been moved to CHA A. When CHA A refreshes they see "Ohh a new report just appeared and I didn't submit this."

I couldn't replicate this. After a login, the places were correct. The only window this scenario could happen is if the CHA gets reassigned but hasn't synced the changes from the server so they still see the CHUs. After a sync/refresh/login the re-assignment is effected correctly.

> UI: The top navbar is missing while on the Reassign Community Health Unit page

It is intentional, only the logo is shown which takes you back to the homepage https://github.com/medic/cht-user-management/issues/269

> It still does not give feedback to users if the contact is not a CHA. If I try to Reassign a CHU to a valid contact who is a CHP, the page refreshes, I don't see the success but at the same time I don't see a fail. Do you think we need to check if the contact has the cha role ? and if they don't then we show an error, the same way we display when contact is not found via an incorrect uuid ?

Done. I've added the check for the roles as a list intersection rather than a 1-1 incase the user gets assigned other roles outside the UMT.

> UI change - The Reassign Community Health Unit page is inconsistent from the other pages like Move Community Health Unit (for instance the header is not bold. It would be nice to have consistent design on the pages)

Fixed
